File: Dynamic Programming/RNA_secondary_structure.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void pr_matrix(int **matrix, int size)
{
		printf("\t\t");
		for (int i = 0; i < size; i++)
		{
				printf("%2d: ", i+1);
				for (int j = 0; j < size; j++)
						printf("%2d ", *(*(matrix + i) + j));
				printf("\n\t\t");
		}
		printf("\n\t\t    ");
		for (int i = 0; i < size; i++)
				printf("%2d ", i+1);
		printf("\n");
}

int ismatch(char a, char b)
{
		if (a == 'A' && b == 'U')
				return(1);
		else if (a == 'U' && b == 'A')
				return(1);
		else if (a == 'G' && b == 'C')
				return(1);
		else if (a == 'C' && b == 'G')
				return(1);
		return(0);
}

int look_up(int **matrix, int start, int stop)
{
		int max = 0, sum;
		for (int k = start + 1; k < stop; k++)
		{
				sum = *(*(matrix + start) + k - 1) + *(*(matrix + k + 1) + stop);
				if (sum > max)
						max = sum;
		}
		return max;
}

int max(int a, int b)
{
		if (a > b)
				return(a);
		return(b);
}

void RNA_secondary_structure(char *str, int size)
{
		int i, j, k, match;
		int **matrix = malloc(sizeof(int *) * size);
		for (i = 0; i < size; i++)
		{
				*(matrix + i) = malloc(sizeof(int) * size);
				memset(*(matrix + i), 0, sizeof(int) * size);
		}
		for (k = 5; k < size; k++)
		{
				for (i = 0; i < size - k; i++)
				{
						j = i + k;
						match = ismatch(str[i], str[j]);
						*(*(matrix + i) + j) = max(*(*(matrix + i) + j - 1), match + look_up(matrix, i, j));
				}
		}
		pr_matrix(matrix, size);
}



int main(int argc, char *argv[])
{
		char *RNA = "ACCGGUAGUCGGAT";
		printf("\t\t%s\n", RNA);
		RNA_secondary_structure(RNA, strlen(RNA));
		return(0);
}

File: Dynamic Programming/RNA_secondary_structure.py
def pr_matrix(matrix):
    print("")
    for i in range(len(matrix)):
        print("\t\t%2d" % (i + 1,), " ", matrix[i])
    print("     \n\t\t", end="")
    for i in range(len(matrix[0])):
        print("%2d " % (i+1,), end="")
    print("\n\t\t", end="")

def pr_result(matrix, string, maxnum):
    def ninset(tup, setA):
        inside = False
        for i in tup:
            if i in setA:
                inside = True
        if not inside:
            return True
        return False
    def tree_recursive(num_list, index, setA, result):
        nonlocal index_list
        if index < 0:
            print("\n\t\t", end="")
            for each_tup in result:
                print("(%d, %d) " % (each_tup[0] + 1, each_tup[1] + 1), end="")
            return
        for each_tup in num_list[index]:
            if ninset(each_tup, setA):
                index_list[index] = 1
                setB = setA | set(each_tup)
                resultB = [i for i in result]
                resultB.append(each_tup)
                tree_recursive(num_list, index - 1, setB, resultB)
            elif (each_tup == num_list[index][-1] and index_list[index] == 0):
                tree_recursive(num_list, index - 1, setA, result)

    if maxnum == 0:
        print("\n\t\tNo results!")
        return
    print("\n\t\tmatch: %s" % (string,), end="")
    num_list = [[] for i in range(maxnum)]
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            current_num = matrix[i][j]
            if current_num != 0:
                num_list[current_num - 1].append((i, j))

    index_list = [0] * len(num_list)
    index_list[-1] = 1
    for tup in num_list[-1]:
        setA = set(tup)
        result = []
        result.append(tup)
        tree_recursive(num_list, len(num_list) - 2, setA, result)



def RNA_match(string):
    def max_t(matrix, start, end):
        maxnum = 0
        for t in range(start + 1, end):
            summ = matrix[start][t - 1] + matrix[t + 1][end]
            if summ > maxnum:
                maxnum = summ
        return maxnum
    def ismatch(a, b):
        if a == "A" and b == "U":
            return 1
        elif a == "U" and b == "A":
            return 1
        elif a == "G" and b == "C":
            return 1
        elif a == "C" and b == "G":
            return 1
        return 0

    length = len(string)
    matrix = []
    maxnum = 0
    for i in range(length):
        matrix.append([0]*length)
    #pr_matrix(matrix)
    for k in range(5, length):
        for i in range(length-k):
            j = i + k
            match = ismatch(string[i], string[j])
            result = max(matrix[i][j - 1], match + max_t(matrix, i, j))
            if result > maxnum:
                maxnum = result
            matrix[i][j] = result
    pr_matrix(matrix)
    pr_result(matrix, string, maxnum)
    print("")


if __name__ == "__main__":
    RNA = "ACCGGUAGUC"
    RNA_match(RNA)
